mypy/dmypy_server.py

```diff
@@ -715,6 +716,29 @@ def refresh_file(module: str, path: str) -> list[str]:
 
         return messages
 
+    def _seen_and_ancestors(self, seen: set[str]) -> set[str]:
+        """Return the set of seen modules along with any ancestors not already in the set.
+
+        For example, given this set:
+
+            {"foo", "foo.bar", "a.b.c"}
+
+        ... we would expect this set to be returned:
+
+            {"foo", "foo.bar", "a.b.c", "a.b", "a"}
+
+        This is used to stop us from deleting ancestor modules from the graph
+        when their descendants have been seen.
+        """
+        seen_paths = seen.copy()
```

Comment:
No big deal, but we could do this without any conditionals like this, if you prefer:
```
    seen_and_ancestors: set[str] = set()
    for module in seen:
        module_parts = module.split(".")
        for i in range(len(module_parts)):
            seen_and_ancestors.add(".".join(module_parts[:i + 1]))
    return seen_and_ancestors
```

Reply:
I assumed that in some cases with many sub-modules that my way would be faster so my preference would be to keep it as-is, but I'm not strongly opposed to changing this.